File: frontend/src/api/simulation.ts
import { apiClient } from './client';
import type { CustomConditions, ScenarioRequest, ScenarioResponse, SimulationStatusOut } from './types';

/**
 * The backend's VALID_SCENARIOS set. UI preset IDs that don't match a native
 * backend scenario are mapped to the closest one (or CUSTOM) so the injection
 * is never rejected with a 422.
 */
export const SCENARIO_NAME_MAP: Record<string, string> = {
  CUSTOM: 'CUSTOM',
  GENERATOR_FAILURE: 'GENERATOR_FAILURE',
  EXTREME_BLIZZARD: 'EXTREME_COLD',
  POLAR_NIGHT_SURGE: 'HIGH_ENERGY_DEMAND',
  CRITICAL_FUEL_SHORTAGE: 'FUEL_SHORTAGE',
  EXTREME_COLD: 'EXTREME_COLD',
  HIGH_ENERGY_DEMAND: 'HIGH_ENERGY_DEMAND',
  FUEL_SHORTAGE: 'FUEL_SHORTAGE',
  EQUIPMENT_DEGRADATION: 'EQUIPMENT_DEGRADATION',
  SUPPLY_DELAY: 'SUPPLY_DELAY',
  NORMAL_OPERATION: 'NORMAL_OPERATION',
};

/** Resolve a UI scenario name to a valid backend scenario name. */
export function resolveScenarioName(name: string): string {
  return SCENARIO_NAME_MAP[name] ?? 'CUSTOM';
}

export const runSimulationScenario = async (
  stationId: number | string,
  scenarioType: string = 'CUSTOM',
  conditions?: CustomConditions,
  applyToLive: boolean = true,
  durationMinutes: number = 60,
  equipmentId?: number
): Promise<ScenarioResponse> => {
  const req: ScenarioRequest = {
    station_id: stationId,
    scenario: resolveScenarioName(scenarioType),
    apply_to_live: applyToLive,
    duration_minutes: durationMinutes,
    equipment_id: equipmentId,
    custom_conditions: conditions,
  };
  const { data } = await apiClient.post<ScenarioResponse>('/simulation/scenario', req);
  return data;
};

export const getSimulationStatus = async (): Promise<SimulationStatusOut> => {
  const { data } = await apiClient.get<SimulationStatusOut>('/simulation/status');
  return data;
};

export const getActiveConditions = async (stationId: number | string) => {
  const { data } = await apiClient.get(`/simulation/active-conditions/${stationId}`);
  return data;
};

export const resetSimulation = async () => {
  const { data } = await apiClient.post('/simulation/reset');
  return data;
};

File: backend/app/api/simulation.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from app.core.database import get_db
from app.schemas.simulation import ScenarioRequest, ScenarioResponse, SimulationStatusOut
from app.services.simulation_service import simulation_service
from app.core.security import APIError

router = APIRouter(prefix="/simulation", tags=["Simulation Engine & What-If Scenarios"])

VALID_SCENARIOS = {
    "NORMAL_OPERATION",
    "GENERATOR_FAILURE",
    "EXTREME_COLD",
    "HIGH_ENERGY_DEMAND",
    "FUEL_SHORTAGE",
    "EQUIPMENT_DEGRADATION",
    "SUPPLY_DELAY",
    "CUSTOM",
}


@router.get("/status", response_model=SimulationStatusOut)
def get_simulation_status():
    """Retrieves current simulation lifecycle status, tick interval, active scenarios, and cycles executed."""
    return simulation_service.get_status()


@router.get("/active-conditions/{station_id}")
def get_active_conditions(station_id: str, db: Session = Depends(get_db)):
    """Retrieves the currently active custom conditions and active scenario for a specific station."""
    from app.services.station_service import station_service
    station = station_service.get_station_by_id_or_code(db, station_id)
    code = station.code.upper()
    return {
        "station_id": station.id,
        "station_code": code,
        "active_scenario": simulation_service.active_scenarios.get(code, "NORMAL_OPERATION"),
        "active_conditions": simulation_service.active_conditions.get(code, None),
        "expires_at": simulation_service.scenario_expiries.get(code, None),
    }


@router.post("/start")
def start_simulation():
    """Resumes or starts the background Digital Twin simulation cycle."""
    running = simulation_service.start()
    return {"success": True, "message": "Digital Twin simulation started.", "is_running": running}


@router.post("/stop")
def stop_simulation():
    """Pauses the background Digital Twin simulation cycle."""
    running = simulation_service.stop()
    return {"success": True, "message": "Digital Twin simulation paused.", "is_running": running}


@router.post("/reset")
def reset_simulation(db: Session = Depends(get_db)):
    """Resets all stations back to standard nominal operation (NORMAL_OPERATION) and clears active failures."""
    res = simulation_service.reset(db)
    return {"success": True, "message": res["message"]}


@router.post("/scenario", response_model=ScenarioResponse)
def trigger_scenario(
    request: ScenarioRequest,
    db: Session = Depends(get_db),
):
    """
    Executes What-If simulation scenarios (e.g. GENERATOR_FAILURE, EXTREME_COLD, HIGH_ENERGY_DEMAND, CUSTOM).
    Calculates immediate system impacts, affected subsystems, and operational recommendations.
    """
    scenario_clean = request.scenario.strip().upper()
    if scenario_clean not in VALID_SCENARIOS:
        raise APIError(
            code="INVALID_SCENARIO",
            message=f"Scenario '{request.scenario}' is not recognized. Valid options: {sorted(list(VALID_SCENARIOS))}",
            status_code=400,
        )

    return simulation_service.apply_scenario(db, request)
